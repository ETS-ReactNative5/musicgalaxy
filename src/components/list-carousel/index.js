export { ListCarousal } from './list-carousel';