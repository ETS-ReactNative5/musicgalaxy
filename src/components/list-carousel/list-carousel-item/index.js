export { ListCarousalItem } from './list-carousel-item';