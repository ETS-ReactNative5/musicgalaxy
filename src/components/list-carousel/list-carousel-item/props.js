import { string, number } from 'prop-types';


export const propTypes = {
    imageSrc: string,
    title: string,
    rating: number,
    releaseDate: string,
    id: number.isRequired
};

export const defaultProps = {
    imageSrc: '',
    title: '',
    rating: 0,
    releaseDate: ''
};
