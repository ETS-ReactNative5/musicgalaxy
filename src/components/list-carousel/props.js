import { array, string, func, bool } from 'prop-types';

export const propTypes = {
    movies: array.isRequired,
    headerTitle: string,
    onPressViewAll: func,
    shouldShowViewAll: bool
};

export const defaultProps = {
    headerTitle: '',
    onPressViewAll: () => { },
    shouldShowViewAll: true
};
