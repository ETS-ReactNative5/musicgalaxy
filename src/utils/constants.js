export const NAVIGATION_ROUTES = {
    HOME: 'Home',
    SEARCH: 'Search',
    ONBOARDING: 'Onboarding',
    VIEW_ALL: 'List',
    SEARCH: 'Search'
};

export const TEST_IDS = {
    FALLBACK_TITLE: 'FALLBACK_TITLE',
    FALLBACK_IMAGE_SRC: 'FALLBACK_IMAGE_SRC',
    LIST_CAROUSEL_ITEM_TITLE: 'LIST_CAROUSEL_ITEM_TITLE',
    LIST_CAROUSEL_ITEM_IMAGE_SRC: 'LIST_CAROUSEL_ITEM_IMAGE_SRC',
    LIST_CAROUSEL_ITEM_ARTIST_TEXT: 'LIST_CAROUSEL_ITEM_ARTIST_TEXT',
    LIST_CAROUSEL_ITEM_RELEASE_DATE: 'LIST_CAROUSEL_ITEM_RELEASE_DATE',
    LIST_CAROUSEL_HEADER_TITLE: 'LIST_CAROUSEL_HEADER_TITLE',
    LIST_CAROUSEL_SEE_ALL_TITLE: 'LIST_CAROUSEL_SEE_ALL_TITLE',
    LIST_CAROUSEL_FLATLIST: 'LIST_CAROUSEL_FLATLIST',
    TEXT_INPUT: 'TEXT_INPUT',
}

export const bannerImgSrc = 'https://images.unsplash.com/photo-1508700115892-45ecd05ae2ad?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=3269&q=80';

export const backIconSvg = `<svg viewBox="0 0 24 24" xmlns="http://www.w3.org/2000/svg">
<path d="M3.5 12L2.79289 11.2929C2.40237 11.6834 2.40237 12.3166 2.79289 12.7071L3.5 12ZM10.2071 17.2929L4.20711 11.2929L2.79289 12.7071L8.79289 18.7071L10.2071 17.2929ZM4.20711 12.7071L10.2071 6.70711L8.79289 5.29289L2.79289 11.2929L4.20711 12.7071ZM3.5 13L20.5 13L20.5 11L3.5 11L3.5 13Z" />
</svg>`;

export const searchIconSvg = `<svg viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg">
<path d="M16 10C16 13.3137 13.3137 16 10 16V18C14.4183 18 18 14.4183 18 10H16ZM10 16C6.68629 16 4 13.3137 4 10H2C2 14.4183 5.58172 18 10 18V16ZM4 10C4 6.68629 6.68629 4 10 4V2C5.58172 2 2 5.58172 2 10H4ZM10 4C13.3137 4 16 6.68629 16 10H18C18 5.58172 14.4183 2 10 2V4ZM14.2421 15.6574L20.5418 21.9571L21.956 20.5429L15.6563 14.2432L14.2421 15.6574Z"/>
</svg>`;
