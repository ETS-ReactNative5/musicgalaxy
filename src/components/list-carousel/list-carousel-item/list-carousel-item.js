import { TEST_IDS } from '@src/utils/constants';
import React, { useState } from 'react';

import {
    StyledImage,
    StyledWrapper,
    MovieInfoWrapper,
    Title,
    MetaInfo,
    Container,
    SubTitle,
} from './styled';

/**
 * Carousel Item component of individual List Item containing info 
 * about the Music Video like Img, Title, Artist etc
 */
export const ListCarousalItem =
    ({
        imageSrc,
        title,
        releaseDate,
        artist,
    }) => {
        const [isLoading, setIsLoading] = useState(true);
        /**
         * Function used to set the image source of item card
         * based on the loading state of image and showing the 
         * loading indicator placeholder instead of actual loader
         * to make the UI responsive  
         */
        const getImageSrc = () => {
            if (isLoading) {
                return require('@assets/loading.png')
            } else {
                return { uri: imageSrc }
            }
        }

        return (
            <Container>
                <StyledWrapper>
                    <StyledImage
                        testID={TEST_IDS.LIST_CAROUSEL_ITEM_IMAGE_SRC}
                        onLoadEnd={() => setIsLoading(false)}
                        source={getImageSrc()}
                    />
                    <MovieInfoWrapper>
                        <Title testID={TEST_IDS.LIST_CAROUSEL_ITEM_TITLE}>
                            {title}
                        </Title>
                    </MovieInfoWrapper>
                </StyledWrapper>
                <MetaInfo>
                    <SubTitle testID={TEST_IDS.LIST_CAROUSEL_ITEM_ARTIST_TEXT}>
                        {artist}
                    </SubTitle>
                    <SubTitle testID={TEST_IDS.LIST_CAROUSEL_ITEM_RELEASE_DATE}>
                        {releaseDate}
                    </SubTitle>
                </MetaInfo>
            </Container>
        )
    };
