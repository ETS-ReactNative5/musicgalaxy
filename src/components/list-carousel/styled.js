import { APP_COLORS } from '@src/theme/colors';
import { spacing } from '@src/theme/spacing';
import { scaleFont } from '@src/utils/media';
import styled from 'styled-components/native';

export const CarouselListWrapper = styled.View``;

export const CarouselListHeader =  styled.Text`
    font-weight: bold;
    font-size: ${scaleFont(25)};
    color: ${APP_COLORS.orange};  
`;

export const HeaderContainer= styled.View`
    flex-direction: row;
    justify-content: space-between;
    margin-vertical: ${spacing.medium}px;
    margin-horizontal: ${spacing.default}px;
    align-items: center;  
`;

export const ViewAll =  styled.Text`
    font-size: ${scaleFont(14)};
    color: #176fb9;
    text-decoration-line: underline;
`;
