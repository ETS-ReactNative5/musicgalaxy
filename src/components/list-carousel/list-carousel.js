import React from 'react';
import { FlatList } from 'react-native';

import { ListCarousalItem } from './list-carousel-item';
import { CarouselListHeader, CarouselListWrapper, HeaderContainer, ViewAll } from './styled';
import { defaultProps, propTypes } from './props';
import { TEST_IDS } from '@src/utils/constants';

/**
 * This component is responsible to render the list of 
 * Music Video Cards and expects the array of movies to render
 * in each carousel based on the Genre of Music
 */
export const ListCarousal =
    ({
        movies,
        headerTitle,
        shouldShowViewAll,
        onPressViewAll
    }) => {

        const renderItem = ({ item }) => (
            <ListCarousalItem
                imageSrc={item.image_url}
                artist={item.artist}
                title={item.title}
                releaseDate={item.release_year}
                id={item.id}
                isFeed
            />
        );

        return (
            <CarouselListWrapper>
                <HeaderContainer>
                    <CarouselListHeader testID={TEST_IDS.LIST_CAROUSEL_HEADER_TITLE} >
                        {headerTitle}
                    </CarouselListHeader>
                    {
                        shouldShowViewAll &&
                        (
                            <ViewAll onPress={onPressViewAll} testID={TEST_IDS.LIST_CAROUSEL_SEE_ALL_TITLE} >
                                See All
                            </ViewAll>
                        )
                    }

                </HeaderContainer>
                <FlatList
                    testID={TEST_IDS.LIST_CAROUSEL_FLATLIST}
                    horizontal
                    data={movies}
                    keyboardShouldPersistTaps='handled'
                    keyExtractor={(item) => `list-carousel-${item.id}`}
                    listKey={Math.random().toString()}
                    renderItem={renderItem}
                />
            </CarouselListWrapper>

        )
    };


ListCarousal.propTypes = propTypes;
ListCarousal.defaultProps = defaultProps;
