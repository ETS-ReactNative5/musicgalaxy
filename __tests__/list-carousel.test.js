import React from 'react';
import { FlatList } from 'react-native';
import { render, fireEvent } from '@testing-library/react-native';
import { NAVIGATION_ROUTES, TEST_IDS } from '@src/utils/constants';
import { ListCarousal } from '@src/components/list-carousel';
import { ListCarousalItem } from '@src/components/list-carousel/list-carousel-item';


const sampleProps = {
    headerTitle: 'Rock',
    movies: [
        {
            artist: "El Koala",
            genre_id: 8,
            id: 501649,
            release_year: 2014,
            title: "Veni paca to"
        },
        {
            artist: "Tom Petty and the Heartbreakers",
            genre_id: 8,
            id: 501895,
            release_year: 2010,
            title: "I Should Have Known It"
        }
    ],
    onPressViewAll: jest.fn,
    shouldShowViewAll: true
};

describe('Testing List Carousel Component', () => {

    it('should match the List Carousel component snapshot', () => {
        const { toJSON } = render(<ListCarousal {...sampleProps} />);
        expect(toJSON()).toMatchSnapshot();
    });

    it('should render List Carousel Header title correctly', () => {
        const { getByTestId } = render(<ListCarousal {...sampleProps} />);
        expect(getByTestId(TEST_IDS.LIST_CAROUSEL_HEADER_TITLE).props.children).toBe('Rock');
    });

    it('should render List Carousel See All correctly', () => {
        const { getByTestId } = render(<ListCarousal {...sampleProps} />);
        expect(getByTestId(TEST_IDS.LIST_CAROUSEL_SEE_ALL_TITLE).props.children).toBe('See All');
    });

    it('should not render List Carousel See All correctly', () => {
        const { queryByTestId } = render(<ListCarousal {...sampleProps} shouldShowViewAll={false} />);
        expect(queryByTestId(TEST_IDS.LIST_CAROUSEL_SEE_ALL_TITLE)).toBeNull();
    });

    it('should navigate to Movie List Screen upon pressing on See All Text', () => {
        const mockedFunc = jest.fn();
        const { getByTestId } = render(<ListCarousal {...sampleProps} onPressViewAll={mockedFunc} />);
        fireEvent(getByTestId(TEST_IDS.LIST_CAROUSEL_SEE_ALL_TITLE), 'press');
        expect(mockedFunc).toHaveBeenCalledWith();
    });

    it('should not render List Carousel Flatlist correctly', () => {
        const { UNSAFE_getAllByType } = render(<ListCarousal {...sampleProps} />);
        expect(UNSAFE_getAllByType(FlatList).length).toBe(1)
    });

    it('should not render List Carousel Flatlist correctly with 2 render Items', () => {
        const { UNSAFE_getAllByType } = render(<ListCarousal {...sampleProps} />);
        expect(UNSAFE_getAllByType(ListCarousalItem).length).toBe(sampleProps.movies.length);
    });

});

