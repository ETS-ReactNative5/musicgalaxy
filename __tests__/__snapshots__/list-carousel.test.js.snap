// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing List Carousel Component should match the List Carousel component snapshot 1`] = `
<View
  style={
    Array [
      Object {},
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "space-between",
          "marginHorizontal": 29,
          "marginVertical": 22,
        },
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#F26E22",
            "fontSize": 25,
            "fontWeight": "bold",
          },
        ]
      }
      testID="LIST_CAROUSEL_HEADER_TITLE"
    >
      Rock
    </Text>
    <Text
      onPress={[Function]}
      style={
        Array [
          Object {
            "color": "#176fb9",
            "fontSize": 15,
            "textDecorationLine": "underline",
          },
        ]
      }
      testID="LIST_CAROUSEL_SEE_ALL_TITLE"
    >
      See All
    </Text>
  </View>
  <RCTScrollView
    data={
      Array [
        Object {
          "artist": "El Koala",
          "genre_id": 8,
          "id": 501649,
          "release_year": 2014,
          "title": "Veni paca to",
        },
        Object {
          "artist": "Tom Petty and the Heartbreakers",
          "genre_id": 8,
          "id": 501895,
          "release_year": 2010,
          "title": "I Should Have Known It",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={true}
    keyExtractor={[Function]}
    keyboardShouldPersistTaps="handled"
    listKey="0.3747739297021"
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    testID="LIST_CAROUSEL_FLATLIST"
    viewabilityConfigCallbackPairs={Array []}
  >
    <View>
      <View
        onLayout={[Function]}
        style={
          Array [
            Object {
              "flexDirection": "row",
            },
            null,
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "marginBottom": 22,
                "marginLeft": 22,
                "marginRight": 22,
                "marginTop": 22,
              },
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                },
              ]
            }
          >
            <View
              style={
                Array [
                  Object {
                    "overflow": "hidden",
                  },
                  Array [
                    Object {
                      "borderTopLeftRadius": 11,
                      "borderTopRightRadius": 11,
                      "height": 326,
                      "resizeMode": "contain",
                      "width": 417,
                    },
                  ],
                ]
              }
            >
              <FastImageView
                onFastImageLoadEnd={[Function]}
                resizeMode="cover"
                source={
                  Object {
                    "testUri": "../../../assets/loading.png",
                  }
                }
                style={
                  Object {
                    "bottom": 0,
                    "left": 0,
                    "position": "absolute",
                    "right": 0,
                    "top": 0,
                  }
                }
                testID="LIST_CAROUSEL_ITEM_IMAGE_SRC"
              />
            </View>
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "backgroundColor": "rgba(12, 12, 12, 0.5)",
                    "borderTopLeftRadius": 30,
                    "bottom": 0,
                    "height": 145,
                    "justifyContent": "center",
                    "paddingBottom": 15,
                    "paddingLeft": 15,
                    "paddingRight": 15,
                    "paddingTop": 15,
                    "position": "absolute",
                    "width": "100%",
                  },
                ]
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "color": "#ffffff",
                      "fontSize": 18,
                      "fontWeight": "bold",
                    },
                  ]
                }
                testID="LIST_CAROUSEL_ITEM_TITLE"
              >
                Veni paca to
              </Text>
            </View>
          </View>
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "backgroundColor": "#27245E",
                  "justifyContent": "center",
                  "paddingVertical": 7,
                },
              ]
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#ffffff",
                    "fontSize": 18,
                    "fontWeight": "bold",
                  },
                ]
              }
              testID="LIST_CAROUSEL_ITEM_ARTIST_TEXT"
            >
              El Koala
            </Text>
            <Text
              style={
                Array [
                  Object {
                    "color": "#ffffff",
                    "fontSize": 18,
                    "fontWeight": "bold",
                  },
                ]
              }
              testID="LIST_CAROUSEL_ITEM_RELEASE_DATE"
            >
              2014
            </Text>
          </View>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={
          Array [
            Object {
              "flexDirection": "row",
            },
            null,
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "marginBottom": 22,
                "marginLeft": 22,
                "marginRight": 22,
                "marginTop": 22,
              },
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                },
              ]
            }
          >
            <View
              style={
                Array [
                  Object {
                    "overflow": "hidden",
                  },
                  Array [
                    Object {
                      "borderTopLeftRadius": 11,
                      "borderTopRightRadius": 11,
                      "height": 326,
                      "resizeMode": "contain",
                      "width": 417,
                    },
                  ],
                ]
              }
            >
              <FastImageView
                onFastImageLoadEnd={[Function]}
                resizeMode="cover"
                source={
                  Object {
                    "testUri": "../../../assets/loading.png",
                  }
                }
                style={
                  Object {
                    "bottom": 0,
                    "left": 0,
                    "position": "absolute",
                    "right": 0,
                    "top": 0,
                  }
                }
                testID="LIST_CAROUSEL_ITEM_IMAGE_SRC"
              />
            </View>
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "backgroundColor": "rgba(12, 12, 12, 0.5)",
                    "borderTopLeftRadius": 30,
                    "bottom": 0,
                    "height": 145,
                    "justifyContent": "center",
                    "paddingBottom": 15,
                    "paddingLeft": 15,
                    "paddingRight": 15,
                    "paddingTop": 15,
                    "position": "absolute",
                    "width": "100%",
                  },
                ]
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "color": "#ffffff",
                      "fontSize": 18,
                      "fontWeight": "bold",
                    },
                  ]
                }
                testID="LIST_CAROUSEL_ITEM_TITLE"
              >
                I Should Have Known It
              </Text>
            </View>
          </View>
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "backgroundColor": "#27245E",
                  "justifyContent": "center",
                  "paddingVertical": 7,
                },
              ]
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#ffffff",
                    "fontSize": 18,
                    "fontWeight": "bold",
                  },
                ]
              }
              testID="LIST_CAROUSEL_ITEM_ARTIST_TEXT"
            >
              Tom Petty and the Heartbreakers
            </Text>
            <Text
              style={
                Array [
                  Object {
                    "color": "#ffffff",
                    "fontSize": 18,
                    "fontWeight": "bold",
                  },
                ]
              }
              testID="LIST_CAROUSEL_ITEM_RELEASE_DATE"
            >
              2010
            </Text>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
